### PR TITLE
Enabled importing ONNX Yolo v3, added Loop op

### PR DESCRIPTION
**Pull request by mbencer**
It is continuation of PR: https://github.com/NervanaSystems/ngraph/pull/4369

### CODE REVIEW

File: ngraph/test/onnx/onnx_import_controlflow.in.cpp [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    EXPECT_EQ(results.size(), 2);
+    EXPECT_EQ(function->get_output_element_type(0), ngraph::element::f32);
+    EXPECT_TRUE(function->get_output_partial_shape(0).is_static());
+    EXPECT_EQ(function->get_output_shape(0), (Shape{1, 2}));
+    EXPECT_EQ(function->get_output_element_type(1), ngraph::element::f32);
+    EXPECT_TRUE(function->get_output_partial_shape(1).is_static());
+    EXPECT_EQ(function->get_output_shape(1), (Shape{3, 2}));
+}
+
+NGRAPH_TEST(${BACKEND_NAME}, onnx_controlflow_loop_2d_add_execution)
```
Comment (mbencer) [inline ngraph/test/onnx/onnx_import_controlflow.in.cpp:180]: I am not able to run Loop tests on backend, it fails in MKLDNNTensorIteratorNode::getSupportedDescriptors() - CVS-33843

File: ngraph/src/ngraph/frontend/onnx_import/op/slice.cpp [hunk 2]
```diff
@@ -196,7 +196,8 @@ namespace ngraph
                                                  std::end(raw_axes_vec),
                                                  [](int64_t axis) { return axis >= 0; }),
                                      "All axes must be positive when data rank is unknown");
-                        axes_vec = std::vector<uint64_t>(std::begin(axes_vec), std::end(axes_vec));
+                        axes_vec =
+                            std::vector<uint64_t>(std::begin(raw_axes_vec), std::end(raw_axes_vec));
```
Comment (tsocha) [inline ngraph/src/ngraph/frontend/onnx_import/op/slice.cpp:200]: Was it a bug?

File: ngraph/src/ngraph/frontend/onnx_import/core/graph.cpp [hunk 3]
```diff
@@ -61,73 +60,35 @@ namespace ngraph
...
+            : Graph(graph_proto, model, std::make_shared<GraphCache>(graph_proto))
+        {
+        }
+
+        Graph::Graph(const ONNX_NAMESPACE::GraphProto& graph_proto,
+                     Model& model,
+                     std::shared_ptr<GraphCache>&& cache)
             : m_graph_proto{&graph_proto}
             , m_model{&model}
+            , m_cache{std::move(cache)}
```
[the remark this answers is not in the document]
Comment (mbencer) [reply ngraph/src/ngraph/frontend/onnx_import/core/graph.cpp]: good point, done

File: ngraph/src/ngraph/frontend/onnx_import/core/graph.cpp [hunk 4]
```diff
@@ -61,73 +60,35 @@ namespace ngraph
...
-                const auto node_name =
-                    onnx_node.get_name().empty() ? "" : onnx_node.get_name() + " ";
-
-                return std::string{"<ONNX " + onnx_node.op_type() + " (" + node_name + "-> " +
-                                   output_names + ")>"};
-            }
         } // namespace detail
 
         Graph::Graph(const ONNX_NAMESPACE::GraphProto& graph_proto, Model& model)
+            : Graph(graph_proto, model, std::unique_ptr<GraphCache>(new GraphCache(graph_proto)))
```
Comment (postrational) [inline ngraph/src/ngraph/frontend/onnx_import/core/graph.cpp:66]: Please use `std::make_unique`

File: ngraph/src/ngraph/frontend/onnx_import/core/graph.cpp [hunk 5]
```diff
@@ -268,6 +226,17 @@ namespace ngraph
...
                 ng_inputs);
         }
+
+        Subgraph::Subgraph(const ONNX_NAMESPACE::GraphProto& proto,
+                           Model& model,
+                           const Graph& parent_graph)
+            : Graph(proto,
+                    model,
+                    std::unique_ptr<SubgraphCache>(
+                        new SubgraphCache(proto, parent_graph.get_graph_cache())))
```
Comment (postrational) [inline ngraph/src/ngraph/frontend/onnx_import/core/graph.cpp:236]: Please use `std::make_unique`

File: ngraph/src/ngraph/op/tensor_iterator.cpp [hunk 6]
```diff
@@ -596,6 +596,18 @@ void op::v0::TensorIterator::validate_and_infer_types()
                 auto axis = concat_output_description->m_axis;
 
                 Shape out_shape{body_value_shape};
+
+                if (body_value_shape.empty())
+                {
+                    NODE_VALIDATION_CHECK(
+                        this,
+                        axis == 0,
+                        "Axis must equal to 0 if concatenated output tensor slices are scalars. "
```
Comment (postrational) [inline ngraph/src/ngraph/op/tensor_iterator.cpp:605]:
```suggestion
                        "Axis must be equal to 0 if concatenated output tensor slices are scalars. "
```

File: ngraph/src/ngraph/frontend/onnx_import/core/graph.cpp (hunk 4 above)
Comment (mbencer) [reply ngraph/src/ngraph/frontend/onnx_import/core/graph.cpp:66]: I would like but it is unfortunately C++14 feature

File: ngraph/src/ngraph/frontend/onnx_import/core/graph.cpp (hunk 5 above)
Comment (mbencer) [reply ngraph/src/ngraph/frontend/onnx_import/core/graph.cpp:236]: as above

File: ngraph/src/ngraph/frontend/onnx_import/op/slice.cpp (hunk 2 above)
Comment (mbencer) [reply ngraph/src/ngraph/frontend/onnx_import/op/slice.cpp:200]: yes

**Comment by mbencer** [comment]
> Just one suggestion to consider - instead of holding and passing `GraphCache` by const pointer, perhaps we could switch to const references.

done

File: ngraph/test/onnx/onnx_import_controlflow.in.cpp [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    try
+    {
+        const auto function = onnx_import::import_onnx_model(file_util::path_join(
+            SERIALIZED_ZOO, "onnx/loop_2d_add_incorrect_access_body_scope.prototxt"));
+        FAIL() << "Incorrect access to body scope not detected";
+    }
+    catch (const ngraph_error& e)
+    {
+        EXPECT_HAS_SUBSTRING(e.what(),
+                             std::string("from_body_scope node not found in graph cache"));
```
[the remark this answers is not in the document]
Comment (mbencer) [reply ngraph/test/onnx/onnx_import_controlflow.in.cpp:109]: It is hard to improve error message here. I added comment with explanation.

**Comment by mbencer** [comment]
> I'm wondering about the access of "_Values from the enclosing scope_". In your implementation the `Subgraph` can reach out only to the initializers from parent graph. What if the loop body would reference to an _inner_ node from the parent graph? I'm not aware of existing models which could have such graph. However at least from programming similarity this is possible situation. Just wanted to catch attention to this.

It works for nodes too. I realized now that pushing initializers to map in GraphCache is not needed

File: ngraph/test/onnx/onnx_import_controlflow.in.cpp [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+    EXPECT_EQ(function->get_output_shape(0), (Shape{}));
+    EXPECT_EQ(function->get_output_element_type(1), ngraph::element::f32);
+    EXPECT_TRUE(function->get_output_partial_shape(1).is_static());
+    EXPECT_EQ(function->get_output_shape(1), (Shape{3}));
+}
+
+NGRAPH_TEST(${BACKEND_NAME}, onnx_controlflow_add_value_from_parent_scope)
+{
+    const auto function = onnx_import::import_onnx_model(
+        file_util::path_join(SERIALIZED_ZOO, "onnx/loop_2d_add_parent_scope.prototxt"));
```
[the remark this answers is not in the document]
Comment (mbencer) [reply ngraph/test/onnx/onnx_import_controlflow.in.cpp]: done - tests for node (output) and input

**Pull request merged**